Archipelago.HollowKnight/Archipelago.cs

```diff
@@ -70,22 +99,57 @@ public override void Initialize(Dictionary<string, Dictionary<string, GameObject
             ModHooks.SavegameLoadHook += ModHooks_SavegameLoadHook;
             ItemChanger.Events.OnItemChangerUnhook += Events_OnItemChangerUnhook;
             ModHooks.HeroUpdateHook += ModHooks_HeroUpdateHook;
-            On.GameCompletionScreen.Start += OnGameComplete;
-
+            // On.GameCompletionScreen.Start += OnGameComplete;
+            ItemChanger.Events.OnSceneChange += Events_OnSceneChange;
+            // ModHooks.SceneChanged += ModHooks_SceneChanged;
             Log("Initialized");
         }
 
-        private void OnGameComplete(On.GameCompletionScreen.orig_Start orig, GameCompletionScreen self)
+        private void Events_OnSceneChange(UnityEngine.SceneManagement.Scene scene)
         {
-            if (ArchipelagoEnabled)
+            string sceneName = scene.name;
+            LogDebug($"Detected change of scene to: {sceneName}.");
+            if (!ArchipelagoEnabled) return;
+
+            Goal goal = SlotOptions.Goal;
+            bool acquiredVoidHeart = PlayerData.instance.GetInt(nameof(PlayerData.royalCharmState)) == 4;
+            bool equippedVoidHeart = acquiredVoidHeart && PlayerData.instance.GetBool(nameof(PlayerData.equippedCharm_36));
+            bool threeDreamers = PlayerData.instance.GetInt(nameof(PlayerData.guardiansDefeated)) >= 3;
+            bool victory = false;
+            LogDebug($"VH Acquired: {acquiredVoidHeart}; VH Equipped: {equippedVoidHeart}; Three Dreamers: {threeDreamers}; Goal: {goal}");
+
+
+            switch (sceneName)
             {
-                session.Socket.SendPacket(new StatusUpdatePacket()
-                {
-                    Status = ArchipelagoClientState.ClientGoal
-                });
+                case SceneNames.Cinematic_Ending_A:  // THK ending
+                    victory = goal == Goal.Any || goal == Goal.HollowKnight;
+                    break;
+                case SceneNames.Cinematic_Ending_B:  // Sealed Siblings ending
+                    victory = goal == Goal.Any || goal == Goal.HollowKnight || goal == Goal.SealedSiblings;
+                    break;
+                case SceneNames.Cinematic_Ending_C:  // Radiance ending
+                    victory = goal == Goal.Any || goal == Goal.HollowKnight || goal == Goal.SealedSiblings || goal == Goal.Radiance;
+                    break;
+                case "Cinematic_Ending_D":  // This isn't listed in SceneNames; it's possible it's unused.  Godhome ending  (TODO: Verify this)
```

Comment:
Are you sure this should fall-through to Cinematic_Ending_E? Maybe this case should simply not be covered.